tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c

```diff
@@ -0,0 +1,278 @@
+		if (!(bytes == validate->exp_value_len &&
+		      memcmp(data, validate->exp_value, bytes) == 0)) {
+			FAIL("Validation failed for key %s", key);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int persistent_values_check(struct key_checker_data *validater)
+{
+
+	key_checked = false;
+	settings_load_subtree_direct("bt", set_load_direct_cb, validater);
+	if (!key_checked) {
+		FAIL("Not found. Key %s", validater->key);
+		return -1;
+	}
+
+	return 0;
+}
+
+/* these data structures are taken from various relavent modules to be checked */
+static struct net_key_val {
+	uint8_t kr_flag : 1,
+		kr_phase : 7;
+	uint8_t val[2][16];
+} __packed;
+
+static struct iv_val {
+	uint32_t iv_index;
+	uint8_t iv_update : 1,
+		iv_duration : 7;
+} __packed;
+
+static struct net_val {
+	uint16_t primary_addr;
+	uint8_t dev_key[16];
+} __packed;
+
+static void test_provisioning_sv_check_save(void)
+{
+	struct key_checker_data validater;
+
+	mount_settings_area();
+	bt_mesh_test_cfg_set(&save_cfg, WAIT_TIME);
+	bt_mesh_test_setup(false);
+
+	/* print raw dump of settings file */
+	char dir_path[100] = CONFIG_SETTINGS_FS_DIR;
+
+	read_dir(dir_path);
+
+	/** check stored provisioning data **/
+	/* check stored netkey matches with expected value */
+	char keypath[20];
+	struct net_key_val keyval = { 0 };
+
+	keyval.kr_flag = 0;
+	keyval.kr_phase = 0;
+	memcpy(keyval.val[0], save_cfg.net_key, sizeof(save_cfg.net_key));
+
+	snprintk(keypath, sizeof(keypath), "mesh/NetKey/%x", save_cfg.net_idx);
+	validater.key = keypath;
+	validater.exp_value = &keyval;
+	validater.exp_value_len = sizeof(keyval);
+
+	persistent_values_check(&validater);
+
+	PASS();
+}
+
+static void test_provisioning_ld_check_load(void)
+{
+	int err;
+
+	mount_settings_area();
+	settings_subsys_init();
+
+	/* provision the device */
+	/* add Netkey with index 0x07, key[16] = {0xAA}, kr=0 */
+	char keypath[20];
+	struct net_key_val key = { 0 };
+
+	key.kr_flag = 0;
+	key.kr_phase = 0;
+	memset(&key.val[0], 0xAA, 16);
+
+	snprintk(keypath, sizeof(keypath), "bt/mesh/NetKey/%x", load_cfg.net_idx);
+	err = settings_save_one(keypath, &key, sizeof(key));
+	if (err) {
+		FAIL("Settings save failed %d", err);
+	}
+
+	/* set IV */
+	struct iv_val iv;
+
+	iv.iv_index = 1111;
+	iv.iv_update = 0;
+	iv.iv_duration = 5;
+	err = settings_save_one("bt/mesh/IV", &iv, sizeof(iv));
+	if (err) {
+		FAIL("Settings save failed %d", err);
+	}
+
+	/* set addr and devkey */
+	struct net_val net;
+
+	net.primary_addr = load_cfg.addr;
+	memcpy(net.dev_key, load_cfg.dev_key, 16);
+	err = settings_save_one("bt/mesh/Net", &net, sizeof(net));
+	if (err) {
+		FAIL("Settings save failed %d", err);
+	}
+
+	/* addr and dev_key are verified indirectly. This should work without fail, as
+	 * it contains additional config client API calls that should work as is
+	 * since device is booting up in already provisioned state.
+	 */
+	bt_mesh_test_cfg_set(&load_cfg, WAIT_TIME);
+	bt_mesh_test_setup(false);
+
+	/* verify IV index state */
```

Comment:
Why is only IV index verified but not the others? I mean you saved a couple of structures above.

Reply:
The addr, dev_key, and net_key are verified indirectly through `bt_mesh_test_setup(false)`. Config client calls won't work if these settings are not loaded properly. I could have added config client calls here but it was looking just a repetition.